Draft: CIP Delegate-less Automation (#1)
julientinguely-da/cips: cip-0058/cip-0058.md

## CIP Delegateless Automation

<pre>
 CIP: N/A
 Title: Delegateless Automation
 Author: Julien Tinguely
 Comments-Summary: No comments yet
 Status: Draft
 Type: Governance
 Created: 2025-04-21
 Approved: N/A
 License: CC0-1.0
</pre>

## Abstract

This CIP proposes to replace the elected SV delegate mechanism to improve fault tolerance. 
Instead of relying on a single delegate, we propose an approach where all SV applications attempt to exercise delegate choices. 
Each application will apply a random delay, tuned to minimize contention.

## Specification

> ### SV Application

Reviewer: I suggest splitting this into a Daml section that describes
1. the new controller argument
2. disabling delegate changes
and an SV app backend changes that describes that each SV attempts to submit with delay and backoff

> The SV application is modified such that it attempts to submit the commands currently only being submitted by the delegate.

Reviewer: ```suggestion
The SV application is modified such that all SVs attempt to submit the commands currently only being submitted by the delegate.
```

In the new implementation, all current choices on the `DsoRules` template having `dsoDelegate` as controller will introduce
a new optional argument named `sv` to be used instead. This allows all SVs to exercise the choice by specifying their own party.
Daml will reject a transaction where they try to specify any party other than their own with an authorization error.

> Once this CIP is implemented, SV applications will try to execute these choices by passing their own SV PartyId as controller.

Reviewer: ```suggestion
```

seems to say nothing that wasn't said before.

This introduces some potential for conflict: if multiple SV applications submit the same transaction at the same time, 
e.g., all attempt to advance the mining rounds at the same time, only one of them will succeed and the others fail, i.e., 
they contend with each other. Importantly, this is not a correctness issue: The Daml code already has to handle this case, 
e.g., if the delegate crashes during a submission and tries to resubmit the same command. But the failed transactions introduce unnecessary load on the network.

To minimize the impact, each SV application adds a random delay before each attempt and retries with a random backoff.

> A new metric named `splice_trigger_attempted_total` is added to monitor trigger contention and chose the right delay and backoff parameters.

Reviewer: ```suggestion
A new metric named `splice_trigger_attempted_total` is added to monitor trigger failures caused by contention and help in choosing good delay and backoff parameters.
```

> ### Remove Delegate Election

Reviewer: ```suggestion
### SV UI
```

In the current implementation, there is a dedicated page in the SV UI to elect a new delegate for the next round, 
intended to unlock the system if a delegate comes down.
In the new implementation, delegate election and its dedicated page in the SV UI are fully removed and 
associated Daml choices cannot be executed anymore.

## Motivation

A delegate performs critical network-related tasks, such as for example advancing mining rounds or executing accepted proposals.
In the current version, there is no automatic fault tolerance to mitigate issues that may occur on the node. 
SVs can vote on a new delegate, but this is a manual process, and the single point of failure is just moved to a different SV.

> A concerning problem observed multiple times is when a malfunctioning delegate causes mining rounds to stop progressing altogether. 

Reviewer: ```suggestion
A concrete problem is rounds stop progressing when the delegate is down.
```
a bit more succinct and neutral

This has been observed multiple times on devnet and testnet and highlights the need for a better fault tolerance mechanism.

## Rationale

All SV applications attempting to complete tasks provides the right fault tolerance guarantees and introduces no correctness issues as commands are already all idempotent. 

> Random delays and backoffs are sufficient to minimize contention given that the delegate based choices do not require high throughput, 

Reviewer: ```suggestion
Given that the current delegate-based choices don't require high throughput, e.g., round advancement happens only every 10 minutes, random delays and backoffs are sufficient to minimize contention.
```

## CIP Delegate-less Automation

<pre>
 CIP: N/A
 Title: Delegate-less Automation
 Author: Julien Tinguely
 Comments-Summary: No comments yet
 Status: Draft
 Type: Governance
 Created: 2025-04-21
 Approved: N/A
 License: CC0-1.0
</pre>

## Abstract

This CIP proposes to replace the elected SV delegate mechanism to improve fault tolerance. 

> Instead of relying on a single delegate, we propose an approach where all SVs attempt to exercise delegate choices. 

Reviewer 2: ```suggestion
Instead of relying on a single delegate, we propose an approach where all SV application attempt to exercise delegate choices. 
```

(to be clear that this is not manual)

> Each applying a random delay tuned to minimize contention.

Reviewer 2: ```suggestion
Each application will apply a random delay, tuned to minimize contention.
```

## Specification

### All SVs try to exercise delegate choices

In the new implementation, all current choices in _dsoRules_ having _dsoDelegate_ as controller will introduce 
a new optional argument named _sv_ to be used instead.
When the new _delegatelessAutomation_ feature flag is set and the _splice-dso-governance.dar_ implementing this change is vetted, 
SV Apps will try to execute these choices by passing their own SV PartyId as controller. 
The main challenge is to avoid contention - concurrent executions of the same choice.

### Minimizing contention

To avoid SVs attempting a task at the same time and having only one succeeding, 
we are adding a random delay before each attempt and tweak the canton retry backoff parameter.
Each random delay D is drawn from a uniform distribution D ~ Uniform(0, U), where the upper bound U is max(n * t, T) 
with n = number of SVs, t = estimated task duration, T = polling trigger interval (30 seconds).

A new metric named `splice_trigger_attempted_total` is added to measure trigger contention.

### Remove Delegate Election

> In the current implementation, there is a dedicated page to elect a new delegate for the next round, 

Reviewer 2: ```suggestion
In the current implementation, there is a dedicated page in the UI to elect a new delegate for the next round, 
```

intended to unlock the system if a delegate comes down.
In the new implementation, delegate election will be fully removed, 
associated Daml choices marked as deprecated and the _Delegate Election_ page in the SV UI removed.

## Motivation

> A delegate performs critical network-related tasks, such as for example advancing mining rounds or close vote proposals. 

Reviewer 2: ```suggestion
A delegate performs critical network-related tasks, such as for example advancing mining rounds or acting on approved vote proposals. 
```

In the current version, many other core processes rely heavily on the delegate, 
with no built-in fault tolerance to mitigate issues that may occur on the node. 
A concerning problem observed multiple times is when a malfunctioning delegate causes mining rounds to stop progressing altogether. 

> This has raised concerns among SV operators and highlights the need for an better fault tolerance mechanism.

Reviewer 2: ```suggestion
This has raised concerns among SV operators and highlights the need for a better fault tolerance mechanism.
```

## Rationale

> ### Requirements

Reviewer 2: I believe the requirements belong in motivation, and then the rationale should explain how the Specification satisfies the requirements.

Author: Thanks for the review Itai! I was not sure how to integrate these requirements. Let me iterate again on that.

Reviewer: I'd honestly just drop the requirements section

Author: dropping

* **Tolerance of faulty SVs:** As long as less than a threshold of SVs is faulty or malicious, 
  all automation should continue to function without significant impact on latency or throughput of tasks processed

* **Minimize Contention:** While some contention are likely unavoidable, it should be minimized. 

>   Given that SVs do not pay traffic contention is not a problem wrt to traffic costs, but it does put additional load on the network.

Reviewer 2: ```suggestion
  Given that SVs do not pay for traffic, contention is not a problem wrt to traffic costs, but it does put additional load on the network.
```

## CIP Delegate-less Automation

<pre>
 CIP: N/A
 Title: Delegate-less Automation
 Author: Julien Tinguely
 Comments-Summary: No comments yet
 Status: Draft
 Type: Governance
 Created: 2025-04-21
 Approved: N/A
 License: CC0-1.0
</pre>

## Abstract

This CIP proposes to replace the elected SV delegate mechanism to improve fault tolerance. 
Instead of relying on a single delegate, we propose an approach where all SV applications attempt to exercise delegate choices. 
Each application will apply a random delay, tuned to minimize contention.

## Specification

### All SVs try to exercise delegate choices

> In the new implementation, all current choices in _dsoRules_ having _dsoDelegate_ as controller will introduce 

Reviewer: ```suggestion
In the new implementation, all current choices on the `DsoRules` template having `dsoDelegate` as controller will introduce 
```

> a new optional argument named _sv_ to be used instead.

Reviewer: ```suggestion
a new optional argument named `sv` to be used instead. This allows all SVs to exercise the choice by specifying their own party. Daml will reject a transaction where they try to specify any party other than their own with an authorization error.
```

> When the new _delegatelessAutomation_ feature flag is set and the _splice-dso-governance.dar_ implementing this change is vetted, 

Reviewer: ```suggestion
Once this CIP is implemented,
```

The feature flag is for us during development, we shouldn't mention that externally.

Author: makes sense

SV applications will try to execute these choices by passing their own SV PartyId as controller. 

> The main challenge is to avoid contention - concurrent executions of the same choice.

Reviewer: ```suggestion
```
the next section describes this already

### Minimizing contention

To avoid SVs attempting a task at the same time and having only one succeeding, 
we are adding a random delay before each attempt and tweak the canton retry backoff parameter.
Each random delay D is drawn from a uniform distribution D ~ Uniform(0, U), where the upper bound U is max(n * t, T) 

> with n = number of SVs, t = estimated task duration, T = polling trigger interval (30 seconds).

Author: missing: should talk about how we tune retries _initialDelay_ and _maxDelay_ so that we don't run into too much contention. Probably increasing the _intialDelay_ is enough here.

Author: Correction: U is max(n * t, T - t)

> A new metric named `splice_trigger_attempted_total` is added to measure and monitor trigger contention.

Reviewer: ```suggestion
### SV App

The sv app is modified such that each SV attempts to submit the commands currently only submitted by the delegate.

This introduces some potential for conflict: If multiple SVs submit the same transaction at the same time, e.g., all attempt to advance the mining rounds at the same time, only one of them will succeed and the others fail, i.e., they contend with each other. Importantly, this is not a correctness issue: The Daml code already has to handle this case, e.g., if the delegate crashes during a submission and tries to resubmit the same command. But the failed transactions introduce unnecessary load on the network.

To minimize the impact, each SV adds a random delay before each attempt and retries with a random backoff.

A new metric named `splice_trigger_attempted_total` is added to measure and monitor trigger contention and chose the right delay and backoff parameters.
```

Reviewer: I think the details of the backoff are not all that interesting for the CIP. The important part is:
1. None of this matters for correctness, it's only about minimizing load and latency.
2. It's monitored and can be tweaked.

### Remove Delegate Election

In the current implementation, there is a dedicated page in the SV UI to elect a new delegate for the next round, 
intended to unlock the system if a delegate comes down.
In the new implementation, delegate election will be fully removed, 

> associated Daml choices marked as deprecated and the _Delegate Election_ page in the SV UI removed.

Reviewer: ```suggestion
associated Daml choices cannot be executed anymore and the _Delegate Election_ page in the SV UI removed.
```

imho the deprecated part is what we do for now during development. Eventually we should just nuke them (i.e. change the implementation to abort)

## Motivation

> A delegate performs critical network-related tasks, such as for example advancing mining rounds or acting on voted proposals. 

Reviewer: ```suggestion
A delegate performs critical network-related tasks, such as for example advancing mining rounds or executing accepted votes.
```

In the current version, many other core processes rely heavily on the delegate, 

> with no built-in fault tolerance to mitigate issues that may occur on the node. 

Reviewer: ```suggestion
In the current version, there is no automatic fault tolerance to mitigate issues that may occur on the node. SVs can vote on a new delegate but this is a manual process and just changes the single point of failure to a different SV.
```

A concerning problem observed multiple times is when a malfunctioning delegate causes mining rounds to stop progressing altogether. 

> This has raised concerns among SV operators and highlights the need for a better fault tolerance mechanism.

Reviewer: ```suggestion
This has been observed several times on devnet and testnet and highlights the need for a better fault tolerance mechanism.
```

### Requirements

* **Tolerance of faulty SVs:** As long as less than a threshold of SVs is faulty or malicious,
  all automation should continue to function without significant impact on latency or throughput of tasks processed

* **Minimize Contention:** While some contention are likely unavoidable, it should be minimized.
  Given that SVs do not pay for traffic, contention is not a problem wrt to traffic costs, but it does put additional load on the network.

* **Accountability of SVs:** It must be possible to detect whether an SV is faulty or malicious or working as expected with high confidence.

* **Well-Behaved under Load:**  Under load where a large number of task piles up, there should be no cascading effects that make things even worse.

## Rationale

### Our new approach

Having all SV applications attempting tasks adds tolerance against faulty and malicious SVs as long as a 2/3 super majority is trustful.
Adding random delays and tweaking backoff retries minimizes contention.
The new metric helps with accountability and tuning our parameters.

> ### Randomization vs. Task-based seeds

Reviewer: I think this section leaks the design doc too much. We don't talk about task based seeds at all here so this title makes no sense.

I would just mention something like

"All SVs attempting to complete tasks provides the right fault tolerance guarantees and introduces no correctness issues as commands are already all idempotent. Random delays and backoffs are sufficient to minimize contention given that the delegate based choices don't require high throughput, e.g., round advancement happens only every 10min.

While more complex approaches to minimize contention would be possible, e.g., defining a clear order in which SVs should attempt to process a given task that is determined based on some task identifier, the implementation complexity of that does not seem to be justified.

Author: ok makes sense again, replacing the current Rationale section with your suggestion.
